File: js-version/factory-helper.js
import * from "./util.js";
import * from "./env.js";

export class FactoryHelper {
	static parseAllEntries(csvString) {
		const NATURE_NAMES = "がんばりや さみしがり ゆうかん いじっぱり やんちゃ ずぶとい すなお のんき わんぱく のうてんき おくびょう せっかち まじめ ようき むじゃき ひかえめ おっとり れいせい てれや うっかりや おだやか おとなしい なまいき しんちょう きまぐれ".split(" ");
		return Util.split(csvString, "\n").map((line, i) => {
			let [pokemon, item, natureName] = line.split(",");
			let nature = NATURE_NAMES.indexOf(natureName);
			return new Entry(i + 1, item, pokemon, nature);
		});
	}
	
	static buildEnv(options) {
		var url = options.allEntriesURL;
		var data;
		await data = Util.xhr(url);
		var allEntries = this.parseAllEntries(data);
		return new Env({
			nParty: options.nParty,
			nStarters: options.nStarters,
			nBattles: options.nBattles,
			allEntries: allEntries
		});
	}
	
	static choose_entry(env, prng) {
		let prngp = prng.dup();
		let x = this.choose_entryQ(env, prngp);
		return [prngp, x];
	}
	
	static choose_entryQ(env, prng) {
		let i = prng.randQ(env.allEntries.length);
		return env.allEntries[i];
	}
	
	static choose_entries(env, prng, n, unchoosable=[]) {
		let prngp = prng.dup();
		let x = this.choose_entriesQ(env, prngp, n, unchoosable);
		return [prngp, x];
	}
	
	static choose_entriesQ(env, prng, n, unchoosable=[]) {
		let entries = [];
		while (entries.length < n) {
			let entry = this.choose_entryQ(env, prng);
			if (!entry.collides_within([...entries, ...unchoosable])) {
				entries.push(entry);
			}
		}
		return entries;
	}

	static _pid_loop(env, prng, entries) {
		for (let entry of entries) {
			let trainer_id = this._rand32Q(prng);
			while (true) {
				let pid = this._rand32Q(prng);
				if (pid % 25 == entry.nature) break;
			}
		}
	}

	static _rand32Q(prng) {
		let low = prng.randQ(0x10000);
		let high = prng.randQ(0x10000);
		return (high << 16 | low) >>> 0;
	}
}

File: ruby-version/factory-helper.rb
Entry = Struct.new(:id, :item, :pokemon)

Entry.class_eval do
  def collides_with?(other)
    self.item == other.item or self.pokemon == other.pokemon
  end

  def collides_within?(entries)
    entries.any?{|x| collides_with?(x) }
  end

  def inspect
    "entry#%03d" % self.id
  end

  def pretty_print(q)
    q.text inspect()
  end
end

class Env
  def initialize(hash)
    @nParty = hash[:nParty]
    @nStarters = hash[:nStarters]
    @nBattles = hash[:nBattles]
    @all_entries = (hash[:all_entries] \
                     or FactoryHelper.load_entries_database(hash[:all_entries_file]))
  end

  attr_reader :nParty, :nStarters, :nBattles, :all_entries
end

module EnvMixin
  def nParty() env.nParty end
  def nStarters() env.nStarters end
  def nBattles() env.nBattles end
end

module FactoryHelper
  module_function
  def load_entries_database(filename)
    open(filename, "rb").lines.map.with_index {|line, i|
      (item, pokemon) = line.split(",").map{|x| Integer(x) }
      Entry.new(i + 1, item, pokemon)
    }
  end

  def gen_all_entries(n_entries, n_items, n_pokemons)
    random = Random.new(0) # 再現性のため種を固定
    (0...n_entries).map {|i|
      item = :"item_#{random.rand(n_items)}"
      pokemon = :"pokemon_#{random.rand(n_pokemons)}"
      Entry.new(i, item, pokemon)
    }
  end

  def collision(a, b)
    a.item == b.item or a.pokemon == b.pokemon
  end

  def choose_entry(env, prng)
    prngp = prng.dup
    x = choose_entry!(env, prngp)
    [prngp, x]
  end

  def choose_entry!(env, prng)
    i = prng.rand!(env.all_entries.size)
    env.all_entries[i]
  end

  def choose_entries(env, prng, n, unchoosable=[])
    prngp = prng.dup
    entries, skipped = choose_entries!(env, prngp, n, unchoosable)
    [prngp, entries, skipped]
  end

  def choose_entries!(env, prng, n, unchoosable=[])
    entries = []
    skipped = []
    while entries.size < n
      entry = choose_entry!(env, prng)
      if entry.collides_within?(entries + unchoosable)
        skipped.push entry
      else
        entries.push entry
      end
    end
    [entries, skipped]
  end
end

